File: testcase_84/services/user_service.go
package services

import (
	"errors"
	"fmt"
	"sync"

	"github.com/myproject/models"
)

// UserService defines the interface for user-related business logic.
type UserService interface {
	GetUserByID(id int) (*models.User, error)
}

// userService implements the UserService interface.
type userService struct {
	// Simulated secure data store (map) for demonstration.
	// In a production environment, this would be a database connection pool.
	users map[int]*models.User
	mu    sync.RWMutex
}

// NewUserService creates a new instance of UserService with mock data.
func NewUserService() UserService {
	// Initialize with mock data
	mockUsers := map[int]*models.User{
		101: {ID: 101, Username: "alice_s", Email: "alice@example.com", IsActive: true},
		102: {ID: 102, Username: "bob_t", Email: "bob@example.com", IsActive: true},
	}
	return &userService{
		users: mockUsers,
	}
}

// GetUserByID retrieves a user by their ID.
// Since the input 'id' is already validated as a positive integer in the handler,
// it is safe to use directly here. In a real DB interaction, this would use
// parameterized queries (prepared statements) to prevent SQL injection.
func (s *userService) GetUserByID(id int) (*models.User, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	user, exists := s.users[id]
	if !exists {
		// Use standard Go errors for internal logic flow (not found).
		return nil, errors.New(fmt.Sprintf("user not found: %d", id))
	}

	return user, nil
}

File: testcase_56/services/user_service.py
from models.user import User
from typing import Optional

# Simulated Database
_MOCK_DB = {
    1: User(id=1, username="alice", email="alice@secure.com"),
    2: User(id=2, username="bob", email="bob@secure.com"),
}

class UserService:
    """
    Handles business logic related to user profiles.
    """

    def get_user_by_id(self, user_id: int) -> Optional[User]:
        """Retrieves a user securely."""
        return _MOCK_DB.get(user_id)

    def update_user_email(self, user_id: int, new_email: str) -> Optional[User]:
        """
        Updates the user's email address in the simulated database.
        
        SECURITY NOTE: In a real application using SQLAlchemy or similar, 
        this operation MUST use parameterized queries (prepared statements) 
        to prevent SQL Injection, even though the input is validated by Pydantic.
        
        Example secure pseudocode for a real DB:
        db.execute("UPDATE users SET email = :new_email WHERE id = :user_id", 
                   {'new_email': new_email, 'user_id': user_id})
        """
        user = self.get_user_by_id(user_id)
        if not user:
            return None

        # Simulate database update
        user.email = new_email
        _MOCK_DB[user_id] = user
        
        return user

user_service = UserService()